### PR TITLE
Factory methods for different IP types.

### PR DESCRIPTION
**Pull request by samuelwilliams**
This addresses issue #57 . Check for a valid string before checking for a packed address.

### CODE REVIEW

**Comment by rlanvin** [comment]
Not a fan of this approach because as you commented in the ticket, this creates the opposite problem without solving the root cause, and the behaviour is still somewhat unpredictable. This will also break backwards compatibility if some people are relying on the first behaviour, so if we're going to make a breaking change, I think there might be more reliable solutions. Let's discuss in #57

**Comment by hwmatthewa** [comment]
I think the only change that I would suggest to this pull request is to retain the current order of checking for valid IPs/Packed strings as @rlanvin mentioned, it would break backwards compatibility to change this.

**Comment by samuelwilliams** [comment]
The current iteration of this PR does not change the order of checking IP strings. It stays the same at:
1) Binary strings
2) Valid IP strings
3) Numeric (decimal) strings

File: src/IP.php [hunk 1]
```diff
@@ -110,6 +110,10 @@ abstract class IP
      */
     public function __construct($ip)
     {
+        if ($ip === null) {
+            return;
```
Comment (rlanvin) [inline src/IP.php]: You are introducing a "backdoor" in the constructor, allowing for construction of invalid objects, and this isn't a good practice. Example:

```
$ip = new PhpIP\IPv4(null);
echo $ip;
```
now completely breaks with unhelpful errors:
> PHP Warning:  gmp_strval(): Unable to convert variable to GMP - wrong type
> PHP Fatal error:  Method PhpIP\IPv4::__toString() must not throw an exception, caught TypeError: Return value of PhpIP\IP::numeric() must be of the type string, bool returned

instead of a clean constructor exception (as currently):
> InvalidArgumentException: Unsupported argument type: "NULL".

File: src/IP.php [hunk 2]
```diff
@@ -138,6 +142,88 @@ public function __construct($ip)
...
 
+    /**
+     * @param int $ipAddress
+     *
+     * @return IP
+     */
+    public static function newFromInteger(int $ipAddress): self
+    {
+        $ip = new static(null);
+        $ip->ip = self::fromInt($ipAddress);
```
Comment (rlanvin) [inline src/IP.php]: Following-up my previous comment: this isn't a good approach, we shouldn't build an invalid object. Instead you could pass the GMP object to the constructor. That would require a bit of refactoring of the constructor to not constantly re-test the type (GMP test could be moved at the very beginning for example), and also maybe moving when isValid is called, so it's only called once in the constructor.

Basically instead of having "conversion+verification" coupled together for each type, we would separate 1. conversion to GMP, and 2. test the GMP object.

**Comment by samuelwilliams** [comment]
I think your observations are accurate. I have implemented the suggested changes.
* The `IP::newFromWhatever()` methods now parse a GMP object directly to the constructor.
* In the constructor, the IP addresses is tested for GMP first.
* `IP::isValid()` is only called once at the end of the constructor.

**Comment by rlanvin** [comment]
Hello @samuelwilliams I know it's been a while since this PR was opened, but I unfortunately didn't have time to look into it earlier. I spent some time playing with it and in the end I made some modifications myself directly. I would be grateful if you could have a look to #61 and tell me what you think of the changes.
